src/operators/products.c

```diff
@@ -0,0 +1,217 @@
+/*
+   IGraph library.
+   Copyright (C) 2025  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_operators.h"
+
+#include "igraph_conversion.h"
+#include "igraph_interface.h"
+
+#include "core/interruption.h"
+#include "math/safe_intop.h"
+
+
+static igraph_error_t cartesian_product(igraph_t *res,
+   const igraph_t *g1, const igraph_t *g2) {
+      
+   igraph_bool_t directed1 = igraph_is_directed(g1);
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Cartesian product of directed and undirected graphs is not supported. "
+                   "You can convert an undirected graph to a directed one using igraph_to_directed, "
+                   "or a directed graph to an undirected one using igraph_to_undirected.", IGRAPH_EINVAL);
```

Comment:
Instead of saying that this is not supported, let's say that the operation is invalid.

Let's not refer to C-specific function names in error messages that will be shown in various non-C contexts (within Python, R, Mathematica).